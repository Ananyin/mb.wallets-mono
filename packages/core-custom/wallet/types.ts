import {
  MaybeArray, SimpleThing, TContext,
  VCV1, VCV1Proof, VCV1Skeleton, VCV1Subject,
  VCV1SubjectMA, VCV1Type, VPV1
} from "@affinidi/vc-common"
import {
  Crypto,
  EntityKey,
  KeyChain,
  KeyChainClaim,
  KeyPair
} from "cryptowrapper/types"


export const BASICWALLET_CREDENTIALS = 'credentials'
export const BASICWALLET_KEYS = 'keys'
export const BASICWALLET_IDENTITY = 'identity'

export const IDTYPE_SUBJECT = 'subject'
export const IDTYPE_SUBJECTS = 'subjects'

export const CREDENTIALS_TYPE_KEY_PERMISSION = 'credentials:typekey:permission'

export const ERROR_WALLET_EXISTS = 'ERROR_WALLET_EXISTS'
export const ERROR_IDENTITY_EXISTS = 'ERROR_IDENTITY_EXISTS'
export const ERROR_NO_STORE_PASSWORD = 'ERROR_NO_STORE_PASSWORD'


export type BasicWalletType = 'credentials' | 'keys' | 'identity' | string
export type IdType = 'subject' | 'subjects'

export type Store = {
  exists: (type: BasicWalletType, alias: string) => Promise<boolean>
  extract: <EntityType extends WalletEntity>(type: BasicWalletType, alias: string, password: string)
    => Promise<BasicWallet<EntityType>>
  store: <EntityType extends WalletEntity>(type: BasicWalletType, wallet: BasicWallet<EntityType>, alias: string, password: string) => Promise<void>
  forget: <EntityType extends WalletEntity>(type: BasicWalletType, alias: string) => Promise<BasicWallet<EntityType>>
}

export type WalletContext = {
  didPrefix: string,
  identityDidPrefix: string,
  personDidPrefix: string,
  organizationDidPrefix: string,
  credentialsDidPrefix: string,
  permissionsDidPrefix: string,
  crypto: Crypto
  getCredentialsType: (key: string) => string
  produceIdentity: (entityKey: EntityKey) => Promise<IdentityEntity>
  getDefaultCredentialsContext: () => TContext
  produceDepersonalizedId: (type: IdType, data: Object, prefixId?: string) => Promise<string>
  producePresentation: (wallet: Wallet, resulsts: CredentialsSearchResult[], holder?: Identity) => Promise<VPV1>
  subjectCompare: <Data extends SimpleThing>(left: VCV1Subject<Data>, right: VCV1Subject<Data>) => Promise<boolean>
}

export type Wallet = {
  exists: boolean

  alias: string

  password?: string

  store: Store

  identity: IdentityWallet

  keys: KeysWallet

  credentials: CredentialsWallet
}

export type CredentialsSearchResult = {
  request: CredentialsRequest,
  results: CredentialsEntity[]
}

export type BasicWalletWrapper = {
  create: (claim?: KeyChainClaim) => Promise<BasicWalletWrapper>
  open: (password: string, safe?: boolean) => Promise<BasicWalletWrapper>
  save: (password?: string) => Promise<void>
  close: () => Promise<BasicWalletWrapper>
  forget: () => Promise<void>
  getOperator: () => WalletOpertor
  getHolder: () => HolderWalletOperator
  getIssuer: () => IssuerWalletOperator
  getVerifier: () => VerifierWalletOperator
} & Wallet

export type WalletOpertor = {
  getSubjectsId: <Data extends SimpleThing>(
    subjects: SubjectClaim<Data>[],
    type: VCV1Type,
    holder: Identity
  ) => Promise<string>
  getSubjectId: <Data extends SimpleThing>(
    subject: SubjectClaim<Data>
  ) => Promise<string>
}

export type HolderWalletOperator = {
  createCredentialsClaim: <Data extends SimpleThing>(
    subjectClaims: SubjectClaim<Data>[],
    type: VCV1Type,
    issuer: Identity,
    holder?: Identity
  ) => Promise<CredentialsClaim<Data>>
  createPresentation: (request: PersentationRequest, holder?: Identity) => Promise<VPV1>
  addCredentails: (credentials: CredentialsEntity, alias?: string, claim?: CredentialsClaim) => Promise<void>
}

export type IssuerWalletOperator = {
  issueCredentials: (credentialsClaim: CredentialsClaim, issuer?: Issuer) => Promise<VCV1>
  requestCredentialsClaim: <Data extends SimpleThing>(
    subjectClaims: SubjectClaim<Data>[],
    types: VCV1Type[],
    holder: Identity,
    issuer?: Identity
  ) => Promise<CredentialsClaimRequest<Data>>
}

export type VerifierWalletOperator = {
  requestCredentials: (
    requests: PersentationRequest[],
    holder?: Identity,
    verifier?: Identity
  ) => Promise<PersentationRequest>
  verifyCredentials: (credentials: VCV1, holder?: Identity) => Promise<void>
  verifyPresentation: (presentation: VPV1, request?: PersentationRequest) => Promise<CredentialsVerificationResult[]>
}

export type CredentialsEntity<
  T extends SimpleThing = SimpleThing,
  Subject extends MaybeArray<VCV1Subject<T>> = VCV1SubjectMA<T>
  > = {} & VCV1<Subject>

export type IdentitySubject = SimpleThing & Identity & {}

export type IdentityEntity = {} & CredentialsEntity<IdentitySubject>

export type WalletEntity = KeyPair | CredentialsEntity | IdentityEntity

export type BasicWallet<EntityType extends WalletEntity> = {
  didIndex: { [k: string]: EntityType },
  aliasIndex: { [k: string]: string[] },
  typeSpecificData?: any
}

export type IdentityWallet = {
  identity: IdentityEntity
} & BasicWallet<IdentityEntity>

export type CredentialsWallet = {
  permissions: string[]
  credentials: string[]
} & BasicWallet<CredentialsEntity>

export type KeysWallet = {
  keyChain: KeyChain
} & BasicWallet<KeyPair>

export type Issuer = {
  key: KeyPair
} & Identity

export type SubjectClaim<Data extends SimpleThing> = VCV1Subject<Data>

export type CredentialsClaim<Data extends SimpleThing = SimpleThing> = {
  issuer?: Identity
  claim: VCV1Skeleton<MaybeArray<VCV1Subject<Data>>>
}

export type CredentialsClaimRequest<Data extends SimpleThing> = {
  holder: Identity
  type: VCV1Type,
  subjectClaims: SubjectClaim<Data>[],
  issuer: Identity
}

export type PersentationRequest = {
  verifier: Identity,
  holder?: Identity,
  requests: CredentialsRequest[],
  proof?: VCV1Proof
}

export type CredentialsRequest = {
  issuer?: Identity[],
  type: VCV1Type
}

export type CredentialsVerificationResult = {
  request: CredentialsRequest,
  result: boolean
}

export type Identity = { id: string }

export type createStore = () => Store