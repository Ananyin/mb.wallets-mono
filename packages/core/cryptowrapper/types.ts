
export type Crypto = {
  generateKeyChain: (keyChainClaim?: KeyChainClaim | KeyChain) => Promise<KeyChain>
  generateKey: (KeyChain: KeyChain, claim: KeyPairClaim) => Promise<KeyPair>
  addEntityKey: (keyChain: KeyChain, type?: string, bip?: string) => Promise<KeyChain>
  addKey: (KeyChain: KeyChain, alias: string, key?: KeyPair | KeyPairClaim) => Promise<KeyChain>
  perRotateEntityKey: (KeyChain: KeyChain, newKey?: KeyPair) => Promise<KeyChain>
  rotateEntityKey: (KeyChain: KeyChain, newKey?: KeyPair) => Promise<KeyChain>

  signWithKey: (key: KeyPair, data: string) => Promise<string>
  verifyWithKey: (key: KeyPair|string, signature: string, data: string) => Promise<boolean>

  encryptWithPassword: (password: string, data: string) => Promise<string>
  decryptWithPassword: (password: string, data: string) => Promise<string>

  produceAsymmetry: (senderKey: KeyPair, recipintKey: KeyPair|string, type?: string) => Promise<Assymetry>
  encryptAsymmetry: (assymetry: Assymetry, data: string) => Promise<string>
  decryptAsymmetry: (recipientKey: KeyPair, data: string) => Promise<[Assymetry, string]>
  rotateAsymmetry: (assymetry: Assymetry) => Promise<Assymetry>

  hash: (data: string, type?: string) => Promise<string>
}

export type CryptoContext = {
}

export type Assymetry = {
  senderKey: KeyPair
  recipientKey: KeyPair
  encryptionKey: KeyPair
  decryptionKey: KeyPair
  details: unknown
  type: string
}

export type KeyChain = {
  intialized: boolean,
  mnemonic?: string
  seed: string
  entityKey: EntityKey
  keys: {[k: string]: KeyPair}
  nextEntityKey: EncryptedKeyPair
}

export type KeyChainClaim = {
  mnemonic?: string
  seed?: string
  empty?: boolean
}

export type EntityKey = {
  bip: string
  address: string
  currentRotation: number
  rotations: EntityKeyRotation[]
  type: string
}

export type EntityKeyRotation = {
  keyPair: KeyPair,
  nextDigest: string,
  date: string
}

export type KeyPairClaim = {
  bip?: string
  type?: string
}

export type KeyPair = {
  bip: string
  type: string
  pk: string
  pubkey: string
}

export type EncryptedKeyPair = {
  encryption: string
} & KeyPair