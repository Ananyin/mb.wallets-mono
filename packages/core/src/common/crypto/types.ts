import { CommonKey } from "common/types/key";

export type CryptoHelper = {
  buildSignSuite: (keyOptions: BuildSignSignatureOptions) => Object

  hash: (data: string) => string

  sign: (data: string, key: string) => string

  getRandomBytes: (size: number) => Promise<Uint8Array>

  verify: (signature: string, data: string, key: string) => boolean

  normalizePassword: (password: string) => Uint8Array

  encrypt: (body: string, password: string) => Promise<string>

  decrypt: (chiper: string, password: string)=> Promise<string>

  makeDerivationPath: (index?: number, change?: number, account?: number, bc?: string) => string

  makeId: (key: string, payload?: string, expand?: boolean) => string

  /**
   * @bug seed param implicitly requires Buffer not Uint8Array!
   */
  getKey: (seed: Uint8Array, derivationPath?: string) => CommonKey & { dp: string }
}

export type BuildSignSignatureOptions = {
  publicKey: string
  privateKey: string
  id: string
  controller: string
}