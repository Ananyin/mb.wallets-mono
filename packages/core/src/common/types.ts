import { TContext } from "@affinidi/vc-common"
import { CryptoHelper } from "metabelarusid-common"
import { KeyChainWrapper } from "keys/types"
import { CommonCredentail, CommonCredentailSubject, CommonSubjectType, CommonType, CommonUnsignedCredential } from "./types/credential"
import { CryptoKey } from "metabelarusid-common"

export type BuildCommonContextMethod = (options: {
  keys: KeyChainWrapper
  crypto: CryptoHelper
}) => Promise<CommonContext>

export type CommonContext = {
  keys: KeyChainWrapper
  crypto: CryptoHelper
  buildCredential: CommonBuildCredentialMethod
  signCredential: CommonSignCredentialMethod
  verifyCredential: CommonVerfiyCredentailMethod
}

export type CommonBuildCredentialMethod = <
  SubjectType extends CommonSubjectType = CommonSubjectType,
  Subject extends CommonCredentailSubject<SubjectType> = CommonCredentailSubject<SubjectType>
  >(options: CommonBuildCredentailOptions<SubjectType>) =>
  Promise<CommonUnsignedCredential<Subject>>

export type CommonContextType = TContext

export type CommonBuildCredentailOptions<
  SubjectType extends CommonSubjectType = CommonSubjectType,
  Subject extends CommonCredentailSubject<SubjectType> = CommonCredentailSubject<SubjectType>
  > = {
    id: string,
    type: CommonType
    holder: string,
    subject: Subject,
    issueanceDate?: string
    context: TContext
  }

export type CommonSignCredentialMethod =
  <
    Subject extends CommonCredentailSubject = CommonCredentailSubject
    >(
    unsingedCredential: CommonUnsignedCredential<Subject>,
    issuer: string,
    key: CryptoKey,
    options?: CommonSignCredentialOptions
  ) =>
    Promise<CommonCredentail<Subject>>

export type CommonSignCredentialOptions = {
  controllerRole?: ControllerRole
  buildProofPurposeOptions?: () => Promise<Object>
}

export type CommonVerfiyCredentailMethod = (credential: CommonCredentail) => Promise<boolean>

export type ControllerRole =
  typeof COMMON_CONTROLLER_ROLE_ISSUER
  | typeof COMMON_CONTROLLER_ROLE_HOLDER

export const COMMON_CONTROLLER_ROLE_ISSUER = 'issuer'
export const COMMON_CONTROLLER_ROLE_HOLDER = 'holder'