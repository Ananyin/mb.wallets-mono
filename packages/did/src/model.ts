import {
  CryptoKey,
  CryptoHelper,
  COMMON_CRYPTO_ERROR_NOPK,
  COMMON_CRYPTO_ERROR_NOPUBKEY,
  COMMON_CRYPTO_ERROR_NOID
} from 'metabelarusid-common'
import {
  DIDDocumnet,
  DIDPURPOSE_VERIFICATION,
  DIDHelper,
  DEFAULT_DID_PREFIX,
  MakeDIDIdOptions,
  DIDDocumentPurpose,
  DIDDocumentPayload,
  didPurposeList,
  DID_ERROR_NOVERIFICATION_METHOD
} from 'types'

export const buildDidHelper =
  (crypto: CryptoHelper, didPrefix = DEFAULT_DID_PREFIX): DIDHelper => {
    const _makeDIDId = (key: CryptoKey, options: MakeDIDIdOptions = {}) => {
      if (!key.id) {
        throw new Error(COMMON_CRYPTO_ERROR_NOID)
      }

      return `did:${didPrefix}:${crypto.makeId(
        key.id,
        options.data && options.hash ? crypto.hash(options.data) : options.data,
        options.expand
      )}`
    }

    const _cleanUpDid = (did: string) => did.split(':')[2]

    const _makeDIDProofSignature = (key: CryptoKey, id: string, nonce: string, purposes: DIDDocumentPurpose[]) => {
      if (!key.pk) {
        throw new Error(COMMON_CRYPTO_ERROR_NOPK)
      }
      return crypto.sign(`${nonce}:${purposes.join('|')}:${id}`, key.pk)
    }

    const _verifyDIDProofSignature = (didDoc: DIDDocumnet, key?: CryptoKey) => {
      if (!didDoc.proof.verificationMethod && !key?.pubKey) {
        throw new Error(DID_ERROR_NOVERIFICATION_METHOD)
      }
      const purposes = didPurposeList.filter(
        purpose => didDoc.hasOwnProperty(purpose)
      )

      key = key || {
        pubKey: <string>didDoc.publicKey.find(
          pubKey => pubKey.id === didDoc.proof.verificationMethod
        )?.publicKeyBase58
      }

      if (!key.pubKey) {
        throw new Error(COMMON_CRYPTO_ERROR_NOPUBKEY)
      }

      return crypto.verify(
        didDoc.proof.signature,
        `${didDoc.proof.nonce}:${purposes.join('|')}:${didDoc.id}`,
        key.pubKey
      )
    }

    return {
      makeDIDId: _makeDIDId,

      makeDIDProofSignature: _makeDIDProofSignature,

      verifyDIDProofSignature: _verifyDIDProofSignature,

      createDID: async (key, options = {}) => {
        if (!key.pubKey) {
          throw new Error(COMMON_CRYPTO_ERROR_NOPUBKEY)
        }
        const id = _makeDIDId(key, options)
        const nonce = crypto.base58().encode(await crypto.getRandomBytes(8))

        let purposes = options.purpose || [DIDPURPOSE_VERIFICATION]
        if (!Array.isArray(purposes)) {
          purposes = [purposes]
        }

        const controller = _makeDIDId(key)

        const didDoc: DIDDocumnet = {
          '@context': [
            'https://w3id.org/security/v2'
          ],
          id,
          ...purposes.reduce((memo: DIDDocumentPayload, purpose) => {
            if (!key.pubKey) {
              throw new Error(COMMON_CRYPTO_ERROR_NOPUBKEY)
            }

            memo[purpose] = [{
              id: `${id}#${purpose}-1`,
              type: 'Secp256k1VerificationKey',
              publicKeyBase58: key.pubKey
            }]

            return memo
          }, {}),
          proof: {
            type: 'Secp256k1Signature',
            controller,
            nonce,
            created: new Date().toISOString(),
            signature: _makeDIDProofSignature(key, id, nonce, purposes),
            verificationMethod: `${controller}#publicKey-1`
          },
          publicKey: [{
            id: `${controller}#publicKey-1`,
            type: 'Secp256k1Signature',
            publicKeyBase58: key.pubKey
          }]
        };

        return {
          ...didDoc,
          proof: didDoc.proof
        }
      }
    }
  }