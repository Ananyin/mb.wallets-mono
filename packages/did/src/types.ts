
export type DIDDocumnet = {
  '@context': string | DIDDocumentContext | DIDDocumentContext[]
  id: string,
  controller?: string,
  verificationMethod?: DIDVerificationMethod[]
  authenitcaion?: (string | DIDAuthentication)[]
  assertionMethod?: (string | DIDAssertion)[]
  keyAgreement?: (string | DIDKeyAgreement)[]
  capabilityInvocation?: (string | DIDCapability)[]
  capabilityDelegation?: (string | DIDDelegation)[]
  service?: DIDService[] 
}

export type DIDDocumentContext = {}

export type DIDVerificationItem = {
  id: string
  type: string
  controller: string
  publicKeyMultibase: string
}

export type DIDVerificationMethod = DIDVerificationItem & {}

export type DIDAuthentication = DIDVerificationItem & {}

export type DIDAssertion = DIDVerificationItem & {}

export type DIDKeyAgreement = DIDVerificationItem & {}

export type DIDCapability = DIDVerificationItem & {}

export type DIDDelegation = DIDVerificationItem & {}

export type DIDService = {
  id: string,
  type: string,
  serviceEndpoint: DIDServiceEndpoint
}

export type DIDServiceEndpoint = string | {[key: string]: string} | string[]