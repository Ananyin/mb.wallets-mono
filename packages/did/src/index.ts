require('dotenv').config()

export * from './types'

export * from './model'
