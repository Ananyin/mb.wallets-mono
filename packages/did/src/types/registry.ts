import { CommonCryptoKey } from "metabelarusid-common"
import { DIDDocumnet, DIDHelper } from "../types"

export type DIDDocumentWrapper = {
  did: DIDDocumnet,
  key?: string,
}

export type DIDRegistry = {
  dids: DIDDocumentWrapper[],
}

export type DIDRegistryWrapper = {
  registry: DIDRegistry,

  lookUpDid: <T extends DIDDocumnet| DIDDocumentWrapper>(did: string, wrapped?: boolean) => Promise<T | undefined>

  extractKey: (did: string) => Promise<CommonCryptoKey>
  addDID: AddDIDMethod
  addPeerDID: AddDIDMethod
  helper(): DIDHelper
}

export type AddDIDMethod = (did: DIDDocumnet, key?: string) => void