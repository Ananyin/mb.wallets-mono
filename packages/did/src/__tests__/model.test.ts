require('dotenv').config()

import { buildDidHelper } from "model"
import { DIDDocumnet } from "types"
import { nodeCryptoHelper } from "../../../common/src"


const testContext: {
  didDoc?: DIDDocumnet
} = {}

describe('DID Helper', () => {
  const didHelper = buildDidHelper(nodeCryptoHelper)

  it('Creates DID Id', async () => {
    const key = await nodeCryptoHelper.getKey(await nodeCryptoHelper.getRandomBytes(32))
    const id = didHelper.makeDIDId(key)
    const described = id.split(':')
    expect(described.length).toBe(3)
    expect(described).toContain('did')
    expect(described).toContain(process.env.DID_PREFIX)
    expect(typeof described[2]).toBe('string')
  })

  it('Creates a DID Document', async () => {
    const key = await nodeCryptoHelper.getKey(await nodeCryptoHelper.getRandomBytes(32))

    const didDoc = await didHelper.createDID(key, {
      data: 'Hello world!',
      hash: true
    })

    testContext['didDoc'] = didDoc
    expect(didDoc).toMatchSnapshot({
      id: expect.any(String),
      verificationMethod: [{
        id: expect.any(String),
        publicKeyBase58: expect.any(String)
      }],
      proof: {
        controller: expect.any(String),
        nonce: expect.any(String),
        created: expect.any(String),
        signature: expect.any(String),
        verificationMethod: expect.any(String),
      },
      publicKey: [{
        id: expect.any(String),
        publicKeyBase58: expect.any(String),
      }]
    })
  })

  it('Verifies DID Document from itself', async () => {
    if (!testContext.didDoc) {
      throw new Error('No DID doc from pervious test')
    }
    const result = didHelper.verifyDID(testContext.didDoc)

    expect(result).toBe(true)
  })
})