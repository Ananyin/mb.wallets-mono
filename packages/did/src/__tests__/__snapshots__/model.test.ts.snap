// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DID Helper Creates a DID Document 1`] = `
Object {
  "@context": Array [
    "https://w3id.org/did/v1",
    "https://w3id.org/security/v2",
  ],
  "assertionMethod": Array [
    Any<String>,
  ],
  "id": Any<String>,
  "proof": Object {
    "controller": Any<String>,
    "created": Any<String>,
    "nonce": Any<String>,
    "signature": Any<String>,
    "type": "EcdsaSecp256k1Signature2019",
    "verificationMethod": Any<String>,
  },
  "publicKey": Array [
    Object {
      "id": Any<String>,
      "publicKeyBase58": Any<String>,
      "type": "EcdsaSecp256k1VerificationKey2019",
    },
  ],
  "verificationMethod": Array [
    Object {
      "controller": Anything,
      "id": Any<String>,
      "publicKeyBase58": Any<String>,
      "type": "EcdsaSecp256k1VerificationKey2019",
    },
  ],
}
`;
